lib/versioncontrol/github/github.go

```diff
@@ -0,0 +1,194 @@
+// `TEST_GITHUB_API=yes`. this will enable some additional tests that are not
+// run as part of normal CI.
+
+var log = logrus.WithFields(logrus.Fields{
+	trace.Component: teleport.ComponentVersionControl,
+})
+
+// defaultHaltingPoint represents the default cutoff for the iterator. this value
+// is expanded/truncated to `<major>.<minor>` so e.g. `v1` would halt iteration on the
+// first `v1.0.X` release observed, and `v1.2.3` would halt iteration on the first
+// `v1.2.X` release observed.
+const defaultHaltingPoint = "v8"
+
+const defaultPageSize = 100
+
+// Release represents a github release.
+type Release struct {
+	// Version is the semver version from the git tag of the release.
+	Version string
+
+	// TODO(fspmarshall): decide on a scheme for embedding additional tags
+	// in our git releases (e.g. security-patch=yes, etc).
+}
+
+// release is the representation of a release returned
+// by the github API.
+type release struct {
+	TagName string `json:"tag_name"`
+}
+
+// getter loads a page of releases. we override the standard page loading logic
+// via this type in order to avoid issues with rate-limiting of the unauthenticated
+// releases API.
+type getter func(n, size int) ([]release, error)
+
+// getPage loads the specified page from the unauthenticated github releases API
+// using the provided http client.
+func getPage(clt http.Client, n, size int) ([]release, error) {
+	if n == 0 {
+		return nil, trace.BadParameter("unspecified page number for teleport/releases (this is a bug)")
+	}
+	if size == 0 {
+		return nil, trace.BadParameter("unspecified page size for teleport/releases (this is a bug)")
+	}
+
+	// perform request
+	res, err := clt.Get(fmt.Sprintf("https://api.github.com/repos/gravitational/teleport/releases?page=%d&per_page=%d", n, size))
+	if err != nil {
+		return nil, trace.Errorf("failed to get teleport/releases: %v", err)
+	}
+
+	body, err := io.ReadAll(res.Body)
+	res.Body.Close()
+
+	if res.StatusCode > 299 {
+		// attempt to extract pretty error message
+		var gherr ghError
+		if json.Unmarshal(body, &gherr) == nil {
+			if gherr.Message != "" {
+				return nil, trace.Errorf("teleport/releases request failed with message: %s (%d)", gherr.Message, res.StatusCode)
+			}
+		}
+		// fallback to status code and quoted body
+		return nil, trace.Errorf("teleport/releases request failed with code=%d, body=%q", res.StatusCode, body)
+	}
+
+	// handle error from rsp body read
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return parsePage(body)
+}
+
+// deserialization helper for github api errors
+type ghError struct {
+	Message string `json:"message"`
+}
+
+func parsePage(page []byte) ([]release, error) {
+	var releases []release
+	if err := json.Unmarshal(page, &releases); err != nil {
+		return nil, trace.Errorf("failed to unmarshal releases page: %v", err)
+	}
+	return releases, nil
+}
+
+// Iterator allows lazy consumption of github release pages. Not safe for
```

Comment:
What happens if there is a concurrent change to the release list on the GitHub side? Will we get duplicates?

Reply:
Seems likely that would be the case.  Added docs indicating this as a possibility.  The systems this will eventually plug are intended to not care about duplicates, but still best to make it explicit.